pg_shard.c

```diff
@@ -258,21 +280,42 @@ PgShardPlanner(Query *query, int cursorOptions, ParamListInfo boundParams)
 		selectFromMultipleShards = SelectFromMultipleShards(query, queryShardList);
 		if (selectFromMultipleShards)
 		{
-			Oid distributedTableId = InvalidOid;
 			Query *localQuery = NULL;
 			List *queryRestrictList = QueryRestrictList(distributedQuery);
 			List *remoteRestrictList = NIL;
 			List *localRestrictList = NIL;
+			bool safeToPushDownAggregate = false;
 
 			/* partition restrictions into remote and local lists */
 			ClassifyRestrictions(queryRestrictList, &remoteRestrictList,
 								 &localRestrictList);
 
+			distributedTableId = ExtractFirstDistributedTableId(distributedQuery);
+			safeToPushDownAggregate = SafeToPushDownAggregate(distributedQuery,
+															  distributedTableId);
+
 			/* build local and distributed query */
-			distributedQuery = RowAndColumnFilterQuery(distributedQuery,
-													   remoteRestrictList,
-													   localRestrictList);
-			localQuery = BuildLocalQuery(query, localRestrictList);
+			distributedQuery = BuildDistributedQuery(distributedQuery,
+													 remoteRestrictList,
+													 localRestrictList,
+													 safeToPushDownAggregate);
+			localQuery = BuildLocalQuery(query, localRestrictList,
+										 safeToPushDownAggregate);
+
+			/*
+			 * distributedPlanTargetList is used for determining intermediate
+			 * temporary table's tuple descriptor in the executer. So, if aggregates
+			 * pushed down, we need to update the list to enable localQuery to be
+			 * executed on the intermediate table.
+			 */
+			if (safeToPushDownAggregate)
+			{
+				distributedPlanTargetList = TargetEntryVarList(localQuery->targetList);
+			}
+			else
```

Comment:
@jasonmp85 Else is required since first assignment is done before BuildDistributedQuery call. Also, first assignment is required for single shard selects.